azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +72,176 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
         """
-        # TODO: Add check for policy file present at /etc/waagent_policy.json.
         # Policy should only be enabled if conf flag is true AND policy file is present.
-        return conf.get_extension_policy_enabled()
+        policy_file_exists = os.path.exists(_CUSTOM_POLICY_PATH)
+        return conf.get_extension_policy_enabled() and policy_file_exists
+
+    @property
+    def policy_enforcement_enabled(self):
+        return self._policy_enforcement_enabled
+
+    def __get_policy(self):
+        """
+        Load policy JSON object from policy file (CUSTOM_POLICY_PATH), return as a dict.
+
+        Note that we should only call this function after validating that CUSTOM_POLICY_PATH exists (this is currently
+        done in __init__).
+        """
+        # TODO: Add schema validation for custom policy file and raise relevant error message to user.
+        with open(_CUSTOM_POLICY_PATH, 'r') as f:        # Open file in read-only mode.
+            self._log_policy_event("Custom policy file found at {0}. Using custom policy.".format(_CUSTOM_POLICY_PATH))
+            try:
+                custom_policy = json.load(f)
+            except ValueError as ex:
+                msg = "policy file does not conform to valid json syntax"
+                raise PolicyInvalidError(msg=msg, inner=ex)
+            except Exception as ex:
+                msg = "unable to load policy file"
+                raise PolicyInvalidError(msg=msg, inner=ex)
+            return self.__parse_policy(custom_policy)
+
+    @staticmethod
+    def __parse_policy(custom_policy):
+        """
+        Return a policy combining the default and custom_policy. Any attributes provided in the custom policy
+        override default values used in the template. If attribute is not provided, default is used.
+
+        The value of the "extensions" attribute is a case-folded dict. CRP allows extensions to be any case, so we use
+        case-folded dict to allow for case-insensitive lookup of individual extension policies.
+
+        The expected custom policy format is:
+         {
+            "policyVersion": str,
+            "extensionPolicies": {
+                "allowListedExtensionsOnly": bool,
+                "signatureRequired": bool,
+                "extensions": {
+                    "<extensionName1>": {
+                        "signatureRequired": bool
+                    }
+                }
+            }
+        }
+        """
+        def __validate_attribute_type(attribute_name, attribute_value, expected_type, parent_attribute_name=None):
+            """
+            Helper method to raise PolicyInvalidError with useful error message.
+            """
+            if not isinstance(attribute_value, expected_type):
+                # Error message should refer to "JSON" not "dict".
+                if expected_type == dict:
+                    expected_type_in_msg = "JSON"
```

Comment:
"object", rather than "JSON"

Reply:
added a __parse_dict method and changed the error message to "object"